### PR TITLE
Applied -Wconversion and Fixed the Warnings

### PR DESCRIPTION
**Pull request by ShikharJ**
@certik @isuruf Pardon me for bothering you with such frugal commit messages, but due to the sheer number of the issued warnings, it was difficult to keep track of fixed files. -Wconversion has been removed in the latest commit, just to test the changes. When the work is close to completion, it will be re-applied.
Relevant: #228

### CODE REVIEW

**Comment by certik** [comment]
Thanks for fixing the warnings!

File: symengine/fields.cpp [hunk 1]
```diff
@@ -731,7 +732,7 @@ GaloisFieldDict::gf_ddf_shoup() const
         return factors;
     GaloisFieldDict f(*this);
     unsigned n = this->degree();
-    auto k = std::ceil(std::sqrt(n / 2));
+    auto k = static_cast<unsigned>(std::sqrt(n / 2) + 1);
```
Comment (isuruf) [inline symengine/fields.cpp]: Can you keep the `std::ceil`?

File: symengine/fields.h [hunk 2]
```diff
@@ -510,7 +510,7 @@ class GaloisFieldDict
 
     unsigned int size() const
```
Comment (isuruf) [inline symengine/fields.h]: This should be changed to `size_t`.

File: symengine/fields.h [hunk 3]
```diff
@@ -522,7 +522,7 @@ class GaloisFieldDict
     {
         if (dict_.empty())
             return 0;
-        return dict_.size() - 1;
+        return static_cast<unsigned>(dict_.size() - 1);
```
Comment (isuruf) [inline symengine/fields.h]: Same here

File: symengine/polys/upolybase.h [hunk 4]
```diff
@@ -307,7 +307,7 @@ class ODictWrapper
 
     unsigned int size() const
```
Comment (isuruf) [inline symengine/polys/upolybase.h]: Same here

File: benchmarks/lwbench.cpp [hunk 5]
```diff
@@ -34,7 +34,9 @@ double A()
     }
     auto t2 = std::chrono::high_resolution_clock::now();
 
-    return std::chrono::duration_cast<std::chrono::nanoseconds>(t2 - t1).count()
+    return static_cast<double>(
+               std::chrono::duration_cast<std::chrono::nanoseconds>(t2 - t1)
+                   .count())
```
Comment (isuruf) [inline benchmarks/lwbench.cpp]: Is this change really necessary?

Comment (ShikharJ) [inline benchmarks/lwbench.cpp]: @isuruf -Wconversion catches the implicit type conversion happening here (since the above expression returns an unsigned int. Hence the change.

**Comment by ShikharJ** [comment]
@isuruf @certik I have tried my best to fix the maximum number of issued warnings. Still, about 3 warnings  are issued in Teuchos library, one in symengine_casts.h and another at symengine_rcp.h, (a total of 5), for which I am unable to find a working fix. -Wconversion has been re-applied though.

**Comment by isuruf** [comment]
Can you post the 5 warnings?

**Comment by ShikharJ** [comment]
@isuruf 
```
/home/shikhar/symengine/symengine/utilities/teuchos/Teuchos_RCPNode.cpp:309:32: warning: conversion to ‘int’ from ‘std::multimap<const void*, {anonymous}::RCPNodeInfo>::size_type {aka long unsigned int}’ may alter its value [-Wconversion]
   return rcp_node_list()->size();
                                ^
```
```
/home/shikhar/symengine/symengine/utilities/teuchos/Teuchos_TestForException.cpp:87:15: warning: conversion to ‘int’ from ‘std::__cxx11::basic_string<char>::size_type {aka long unsigned int}’ may alter its value [-Wconversion]
   break_on_me = errorMsg.length(); // Use errMsg to avoid compiler warning.
               ^
```
```
/home/shikhar/symengine/symengine/utilities/teuchos/Teuchos_stacktrace.cpp:383:41: warning: conversion to ‘int’ from ‘std::vector<long long unsigned int>::size_type {aka long unsigned int}’ may alter its value [-Wconversion]
     return this->stacktrace_buffer.size();
                                         ^
```
```
/home/shikhar/symengine/symengine/symengine_rcp.h:380:19: warning: conversion to ‘unsigned int’ from ‘long int’ may alter its value [-Wconversion]
     RCP<T> p = rcp(new T(std::forward<Args>(args)...));
                   ^
```
```
/home/shikhar/symengine/symengine/symengine_casts.h:36:12: warning: conversion to ‘unsigned int’ from ‘double’ may alter its value [-Wfloat-conversion]
     return f;
            ^
```

File: symengine/symengine_casts.h [hunk 6]
```diff
@@ -33,7 +33,7 @@ struct remove_reference<T &> {
 template <typename To, typename From>
 inline To implicit_cast(const From &f)
 {
-    return f;
+    return static_cast<To>(f);
```
Comment (certik) [inline symengine/symengine_casts.h]: I don't think that's the correct fix. I think this should be left as it was, to allow the compiler to check types.

File: symengine/utilities/teuchos/Teuchos_stacktrace.cpp [hunk 7]
```diff
@@ -380,7 +380,7 @@ class StacktraceAddresses {
     return this->stacktrace_buffer[i];
   }
   int get_size() const {
-    return this->stacktrace_buffer.size();
+    return static_cast<unsigned>(this->stacktrace_buffer.size());
```
Comment (certik) [inline symengine/utilities/teuchos/Teuchos_stacktrace.cpp]: Here I think the `get_size` should return `size_t` argument.

File: symengine/utilities/teuchos/Teuchos_stacktrace.cpp [hunk 8]
```diff
@@ -465,7 +465,7 @@ RCP<StacktraceAddresses> get_stacktrace_addresses(int impl_stacktrace_depth)
 #else
   const size_t stacktrace_size = 0;
 #endif
-  return rcp(new StacktraceAddresses(stacktrace_array, stacktrace_size,
+  return rcp(new StacktraceAddresses(stacktrace_array, static_cast<unsigned>(stacktrace_size),
```
Comment (certik) [inline symengine/utilities/teuchos/Teuchos_stacktrace.cpp]: The argument of `StacktraceAddresses` should accept `size_t` instead.

File: symengine/sparse_matrix.cpp [hunk 9]
```diff
@@ -323,7 +323,7 @@ CSRMatrix CSRMatrix::from_coo(unsigned row, unsigned col,
                               const std::vector<unsigned> &j,
                               const vec_basic &x)
 {
-    unsigned nnz = x.size();
+    unsigned nnz = static_cast<unsigned>(x.size());
```
Comment (certik) [inline symengine/sparse_matrix.cpp]: Use `size_t` instead of `unsigned`?

Comment (isuruf) [inline symengine/sparse_matrix.cpp]: This is actually okay since the row indices stored in `CSRMatrix` are unsigned.

**Comment by ShikharJ** [comment]
@certik Indeed, `static_cast` is not the way to go. The point you mentioned (of silent failures) is one of the reasons why many C++ development authors explicitly suggest people to minimise casting. However, the reason I resorted to casting was firstly to signify places that needed improvements. Secondly, I had my doubts about whether a particular "up-gradation" (say changing the return type from int to long) or "down-gradation" (changing from size_t to unsigned) should be applied at a place or not. Instead of clarifying them one by one, I thought it was better to have them reviewed all at once, so that the changes can be applied where they are utmost necessary.

**Comment by isuruf** [comment]
See my changes

**Comment by certik** [comment]
It looks like primesieve has these conversion issues as well.

**Comment by isuruf** [comment]
@ShikharJ, ping

File: symengine/series_flint.cpp [hunk 10]
```diff
@@ -138,8 +138,8 @@ unsigned URatPSeriesFlint::ldegree(const fqp_t &s)
 {
     long i = 0;
     while (i <= s.degree())
-        if (not s.get_coeff(i++).is_zero())
-            return i - 1;
+        if (not((s.get_coeff(i++)).is_zero()))
```
Comment (ShikharJ) [inline symengine/series_flint.cpp]: @isuruf This line is giving away this error:
```
error: conversion to ‘unsigned int’ from ‘long int’ may alter its value [-Werror=conversion]
```
even though `is_zero()` would return a `bool`. Any idea why is this cropping up?

Comment (isuruf) [inline symengine/series_flint.cpp]: Is `.get_coeff` expecting a long?

Comment (ShikharJ) [inline symengine/series_flint.cpp]: I didn't notice that it was. I'll update this. BTW, can you also enable the auto-cancellation of Travis Builds?

File: symengine/polys/basic_conversions.h [hunk 11]
```diff
@@ -143,7 +144,8 @@ class BasicToUPolyBase : public BaseVisitor<V>
         if (eq(*genbase, x)) {
             powr = div(one, genpow);
             if (is_a<const Integer>(*powr)) {
-                int i = down_cast<const Integer &>(*powr).as_int();
+                int i = static_cast<int>(
```
Comment (isuruf) [inline symengine/polys/basic_conversions.h]: Don't use `static_cast` when you don't know whether it will fit or not

File: symengine/polys/uintpoly_flint.h [hunk 12]
```diff
@@ -125,9 +125,9 @@ class UFlintPoly : public BaseType<Container, Poly>
         return r_iterator(this->template rcp_from_this_cast<Poly>(), -1);
     }
 
-    unsigned int size() const
+    int size() const
```
Comment (isuruf) [inline symengine/polys/uintpoly_flint.h]: Can this be negative?

Comment (ShikharJ) [inline symengine/polys/uintpoly_flint.h]: @isuruf I don't think it can. Initially, it was giving an error such as this:
```
symengine/symengine/polys/upolybase.h:377:17: error:   overriding ‘int SymEngine::UPolyBase<Container, Poly>::size() const [with Container = SymEngine::fmpq_poly_wrapper; Poly = SymEngine::URatPolyFlint]’
```
It didn't make much sense to change the parent function. Hence it was changed here. What should be the approach here?

Comment (isuruf) [inline symengine/polys/uintpoly_flint.h]: I don't understand the purpose of having two functions `degree` and `size`, but if it works, okay. cc @srajangarg

Comment (srajangarg) [inline symengine/polys/uintpoly_flint.h]:
```
size() :
    if p == 0 :
        return 0
    else :
        return degree() + 1
```

Comment (ShikharJ) [inline symengine/polys/uintpoly_flint.h]: @srajangarg @isuruf 
Seems like that. The change was made in this [commit](https://github.com/symengine/symengine/pull/1147/commits/a482c0815592008206485d27c49d32c330d0afbf) in `upolybase.h`. I've tried reverting this back, but then we'll have to make even more changes (currently, we just have to account for two `size()` functions, one in  `uintpoly_flint.h` and other in `uintpoly_piranha.h`).

**Comment by ShikharJ** [comment]
@isuruf This is ready for a review.

File: symengine/polys/uintpoly_flint.h (hunk 12 above)
Comment (isuruf) [inline symengine/polys/uintpoly_flint.h]: @srajangarg, yes I know, but what is the use of having the two functions?

**Comment by isuruf** [comment]
Looks good to me, @certik, can you review?

File: symengine/polys/uintpoly_flint.h (hunk 12 above)
Comment (srajangarg) [inline symengine/polys/uintpoly_flint.h]: Flint has both the functions, so I thought it was standard to provide both.

**Comment by ShikharJ** [comment]
Ping @certik.

**Comment by ShikharJ** [comment]
Ping @certik.

**Comment by ShikharJ** [comment]
Ping @certik.

**Comment by ShikharJ** [comment]
Ping @certik.

**Comment by certik** [comment]
The changes look good. Can we enable the `-Wconversion` on Travis and make it an error?

**Comment by certik** [comment]
Are benchmarks affected?

File: symengine/symengine_casts.h [hunk 13]
```diff
@@ -87,6 +90,51 @@ inline To down_cast(From &f)
...
+numeric_cast(From f,
+             typename std::enable_if<(std::is_signed<From>::value
+                                      && std::is_signed<To>::value)
+                                     || (std::is_unsigned<From>::value
+                                         && std::is_unsigned<To>::value)>::type
+                 * = nullptr)
+{
+    SYMENGINE_ASSERT(f <= std::numeric_limits<To>::max());
+    SYMENGINE_ASSERT(f >= std::numeric_limits<To>::min());
+    return static_cast<To>(f);
```
Comment (isuruf) [inline symengine/symengine_casts.h]: @certik, in Release mode `numeric_cast` should have the same cost as `static_cast` (which is zero)

File: symengine/CMakeLists.txt [hunk 14]
```diff
@@ -185,6 +185,11 @@ if (BUILD_SHARED_LIBS)
         SOVERSION ${SYMENGINE_MAJOR_VERSION}.${SYMENGINE_MINOR_VERSION}
     )
 endif()
+
+if (CMAKE_CXX_COMPILER_ID MATCHES "GNU|Clang")
+    set_target_properties(symengine PROPERTIES COMPILE_FLAGS "-Wconversion -Wno-sign-conversion")
```
Comment (isuruf) [inline symengine/CMakeLists.txt]: This warning is enabled by default and in travis-ci all warnings are converted to errors

**Comment by isuruf** [comment]
@ShikharJ, thanks for the PR
